banzai/qc/pattern_noise.py

```diff
@@ -21,49 +22,197 @@ def group_by_keywords(self):
 
+    Parameters
+    ----------
+    data : numpy array
+        The data for computing the Fourier Transform
+    fractional_band_width : float
+        Vertical band width as a fraction of ny
+
+    Returns
+    -------
+    power_2d : numpy array
+        Central band of 2d Fourier transform
+    """
+    # Get full 2D power
+    full_power_2d = abs(np.fft.rfft2(data))
+
+    # Extract horizontal band, as corners of 2D FFT can vary significantly between images
+    ny = full_power_2d.shape[0]
+    y1 = int(ny * (0.5 - fractional_band_width/2))
+    y2 = int(ny * (0.5 + fractional_band_width/2))
+
+    return full_power_2d[y1:y2]
+
+
+def compute_snr(power_2d, fractional_window_size=0.05):
+    """
+    Extract the central region of the 2D Fourier transform
+
+    Parameters
+    ----------
+    power_2d : numpy array
+        The 2D Fourier transform of the data
+    fractional_window_size : float
+        Median filter window size as a fraction of the 1D power array
+
+    Returns
+    -------
+    snr : numpy array
+        The 1D SNR
+    """
+    power = np.median(power_2d, axis=0)
+    p2p_scatter = abs(power[1:] - power[:-1])
+    power = power[1:]  # Throw away DC term
+
+    # Median filter
+    window_size = get_odd_integer(fractional_window_size * len(power))
+    continuum = median_filter(power, size=window_size)
+    scatter = median_filter(p2p_scatter, size=window_size)
+
+    snr = (power - continuum) / scatter
+    return snr
+
+
+def get_odd_integer(x):
+    """
+    Return the closest odd integer given a float
+
+    Parameters
+    ----------
+    x : float
+        The number to be converted to the closest odd integer
+
+    Returns
+    -------
+    y : int
+        Odd integer of x
+    """
+    return int(round(round(x) / 2) * 2) + 1
+
+
+def convolve_snr_with_wavelet(snr,
+                              fractional_wavelet_width_min=0.01,
+                              fractional_wavelet_width_max=0.025,
+                              nwavelets=25):
+    """
+    Convolve the 1D SNR with nwavelets Ricker wavelets, with widths ranging from
+    fractional_wavelength_width_min * len(snr) to fractional_wavelength_witdh_max * len(snr).
+
+    Parameters
+    ----------
+    snr : numpy array
+        The 1D SNR array
+    fractional_wavelet_width_min : float
+        The fraction of the SNR length to use as the minimum wavelet width
+    fractional_wavelet_width_max : float
+        The fraction of the SNR length to use as the maximum wavelet width
+    nwavelets : int
+        The number of wavelength widths to convolve with the SNR
+
+    Returns
+    -------
+    snr_convolved : numpy array
+        Array of size nwavelets x len(snr) of snr convolved with different wavelet widths
+    """
+
+    widths = np.linspace(len(snr) * fractional_wavelet_width_min,
+                         len(snr) * fractional_wavelet_width_max,
+                         nwavelets)
+    return cwt(snr, ricker, widths)
```

Comment:
This might be more readable as scipy.signal.cwt

Reply:
Changed